modules/appgw/variables.tf

```diff
@@ -1,169 +1,632 @@
+  pinned to a single zone or zone-redundant (so available in all zones in a region).
+  Therefore make sure that if you specify more than one zone you specify all available in a region. You can use a subset,
+  but the Public IP will be created in all zones anyway. This fact will cause terraform to recreate the IP resource during
+  next `terraform apply` as there will be difference between the state and the actual configuration.
 
-  For details on zones currently available in a region of your choice refer to [Microsoft's documentation](https://docs.microsoft.com/en-us/azure/availability-zones/az-region).
+  For details on zones currently available in a region of your choice refer to
+  [Microsoft's documentation](https://docs.microsoft.com/en-us/azure/availability-zones/az-region).
 
   Example:
   ```
   zones = ["1","2","3"]
   ```
   EOF
-  default     = null
+  default     = ["1", "2", "3"]
   type        = list(string)
+  validation {
+    condition     = var.zones == null || length(setsubtract(var.zones, ["1", "2", "3"])) == 0
+    error_message = "The `var.zones` can either bea non empty list of Availability Zones or explicit `null`."
+  }
 }
 
-variable "name" {
-  description = "Name of the Application Gateway."
+variable "public_ip_name" {
+  description = "Name for the public IP address."
   type        = string
 }
 
 variable "domain_name_label" {
-  description = "Label for the Domain Name. Will be used to make up the FQDN. If a domain name label is specified, an A DNS record is created for the public IP in the Microsoft Azure DNS system."
+  description = <<-EOF
+  Label for the Domain Name.
+
+  Will be used to make up the FQDN. If a domain name label is specified, an A DNS record is created
+  for the public IP in the Microsoft Azure DNS system."
+  EOF
   default     = null
   type        = string
 }
 
+variable "enable_http2" {
+  description = "Enable HTTP2 on the Application Gateway."
+  default     = false
+  nullable    = false
+  type        = bool
+}
+
+variable "waf" {
+  description = <<-EOF
+  Object sets only the SKU and provide basic WAF configuration for Application Gateway.
+
+  This module does not support WAF rules configuration and advanced WAF settings.
+  Only below attributes are allowed:
+  - `enabled`          - (`bool`, required) Enables WAF Application Gateway
+  - `firewall_mode`    - (`string`, optional) The Web Application Firewall Mode
+  - `rule_set_type`    - (`string`, optional, defaults to `OWASP`) The Type of the Rule Set used for this Web Application Firewall
+  - `rule_set_version` - (`string`, optional) The Version of the Rule Set used for this Web Application Firewall
+  EOF
+  default = {
+    enabled = false
+  }
+  nullable = false
+  type = object({
+    enabled          = bool
+    firewall_mode    = optional(string)
+    rule_set_type    = optional(string, "OWASP")
+    rule_set_version = optional(string)
+  })
+  validation {
```

Comment:
just an idea, does not have to be a good one: maybe instead of adding `enabled` control WAF by either setting this map to `null` (default, means disabled) or by specifying a map (means enabled)

Reply:
I was thinking about that approach, but if we allow to have by default `null` for variable `waf`, then we have problem with conditions e.g. :

```
│ Error: Attempt to get attribute from null value
│ 
│   on ../../modules/appgw/variables.tf line 100, in variable "waf":
│  100:     condition     = var.waf == null || var.waf != null && contains(["OWASP", "Microsoft_BotManagerRuleSet"], var.waf.rule_set_type)
│     ├────────────────
│     │ var.waf is null
│ 
│ This value is null, so it does not have any attributes.
```

By default WAF is disabled and that value: 

```
  default = {
    enabled = false
  }
```

is self-explanatory, very clear to understand and clean to use e.g.:

```
name     = var.waf.enabled ? "WAF_v2" : "Standard_v2"
```

rather that checking if something is `null` or not , then it means that it's enabled or not. 

Can we leave it as it is ?